src/link/Wasm.zig

```diff
@@ -29,6 +30,58 @@ pub const FnData = struct {
     idx_refs: std.ArrayListUnmanaged(struct { offset: u32, decl: *Module.Decl }) = .{},
 };
 
+/// Data section of the wasm binary
+/// Each declaration will have its own 'data_segment' within the section
+/// where the offset is calculated using the previous segments and the content length
+/// of the data
+pub const DataSection = struct {
+    /// Every data object will be appended to this list,
+    /// containing its `Decl`, the data in bytes, and its length.
+    segments: std.ArrayListUnmanaged(struct {
+        /// The decl that lives inside the 'data' section such as an array
+        decl: *Module.Decl,
+        /// The contents of the data in bytes
+        data: [*]const u8,
+        /// The length of the contents inside the 'data' section
+        len: u32,
+    }) = .{},
+
+    /// Returns the offset into the data segment based on a given `Decl`
+    pub fn offset(self: DataSection, decl: *const Module.Decl) u32 {
+        var cur_offset: u32 = 0;
+        return for (self.segments.items) |entry| {
+            if (entry.decl == decl) break cur_offset;
+            cur_offset += entry.len;
+        } else unreachable; // offset() called on declaration that does not live inside 'data' section
```

Comment:
Let's chat about this. I think there is a better solution than linear scan across all decls for this.